### ISSUE TITLE
Type instability in eigs

### ISSUE THREAD

**Issue by dgleich** [JuliaLang/julia#13929]
There seems to be an issue with the eigs (sparse eigenvalue) solver that causes a type instability in the output even for Symmetric inputs.

```
n = 500; A = sparse(1:n-1,2:n,1.,n,n); A = A + A'
@code_warntype eigs(A)
@code_warntype eigs(Symmetric(A))
```

The output I get on today's master is

```
# Version 0.5.0-dev+1195 (2015-11-09 14:48 UTC)

julia> @code_warntype eigs(A)
Variables:
  A::SparseMatrixCSC{Float64,Int64}

Body:
  begin 
      $(Expr(:line, 50, symbol("linalg/arnoldi.jl"), symbol("")))
      GenSym(8) = (top(ccall))(:jl_alloc_array_1d,(top(apply_type))(Base.Array,Any,1)::Type{Array{Any,1}},(top(svec))(Base.Any,Base.Int)::SimpleVector,Array{Any,1},0,0,0)::Array{Any,1}
      GenSym(9) = GenSym(8)
      return (Base.LinAlg.__eigs#222__)(GenSym(9),A::SparseMatrixCSC{Float64,Int64})::Tuple
  end::Tuple

julia> @code_warntype eigs(Symmetric(A))
Variables:
  A::Symmetric{Float64,SparseMatrixCSC{Float64,Int64}}

Body:
  begin 
      $(Expr(:line, 50, symbol("linalg/arnoldi.jl"), symbol("")))
      GenSym(8) = (top(ccall))(:jl_alloc_array_1d,(top(apply_type))(Base.Array,Any,1)::Type{Array{Any,1}},(top(svec))(Base.Any,Base.Int)::SimpleVector,Array{Any,1},0,0,0)::Array{Any,1}
      GenSym(9) = GenSym(8)
      return (Base.LinAlg.__eigs#222__)(GenSym(9),A::Symmetric{Float64,SparseMatrixCSC{Float64,Int64}})::Tuple
  end::Tuple
```

At least in the second case, it'd be really nice to be able to type-stable outputs.

This seems to mirror that seen in the eig function. #12304

References:
julia-users discussion: https://groups.google.com/forum/#!topic/julia-users/Frmt74TgbAs

**Comment by andreasnoack** [comment]
I agree that it would be nice if `eigs(Symmetric(A))` was type stable, but `eigs` will be much harder to fix than #12304. I think it requires a complete restructuring of our ARPACK wrappers.

**Comment by dgleich** [comment]
Are there any tips on debugging these things? I'm willing to take a look and have some experience with calling ARPACK directly from almost a decade ago. (More concretely, what's the fastest way of editing base/linalg/arnoldi.jl and getting julia to see the changes, running make seems to take forever.)

**Comment by andreasnoack** [comment]
Copy `_eigs`, `aupd_wrapper`, and `eupd_wrapper` to a new file and wrap them in a module, say `TEST`, and correct the code such that your versions of the functions are called. Then call @code_warntype on `_eigs` with two arguments, like
```jl

julia> @code_warntype TEST._eigs(Symmetric(A), I)
Variables:
  A::Symmetric{Float64,SparseMatrixCSC{Float64,Int64}}
  B::UniformScaling{Int64}
  nev::Int64
  ncv::Int64
  which::Symbol
  tol::Float64
  maxiter::Int64
  sigma::Void
  v0::Array{Float64,1}
  ritzvec::Bool
  ##dims#7233::Tuple{Tuple{Int64}}

Body:
  begin
      $(Expr(:line, 591, symbol("/Users/andreasnoack/Desktop/test.jl"), symbol("")))
      nev = 6
      GenSym(0) = (Base.box)(Base.Int,(Base.add_int)((Base.box)(Int64,(Base.mul_int)(2,ne
v::Int64)::Any)::Int64,1)::Any)::Int64
      ncv = ((top(getfield))(Base.Intrinsics,:select_value)::I)((Base.slt_int)(GenSym(0),
20)::Bool,20,GenSym(0))::Int64
      which = :LM
      tol = 0.0
      maxiter = 300
      sigma = TEST.nothing
      v0 = (Base.fill!)((Base.Array)(Float64,(top(tuple))(0)::Tuple{Int64})::Array{Float6
4,1},(Base.box)(Float64,(Base.sitofp)(Float64,0)::Any)::Float64)::Array{Float64,1}
      ritzvec = true
      return (TEST.___eigs#3__)(nev::Int64,ncv::Int64,which::Symbol,tol::Float64,maxiter:
:Int64,sigma::Void,v0::Array{Float64,1},ritzvec::Bool,A::Symmetric{Float64,SparseMatrixCS
C{Float64,Int64}},B::UniformScaling{Int64})::Tuple
  end::Tuple
```
This is not the actual function because the function has keyword arguments. The actual name can be after the `return` in the last line so next step is to call `@code_warntype` on the, i.e.
```jl
julia> @code_warntype TEST.(symbol("___eigs#3__"))(6, 20, :LM, 0.0, 300, nothing, zeros(0
), true, Symmetric(A), I)
```
and continue from there.

I tried to see if this could be fixed easily easlier today and I don't think the problem is to understand ARPACK, but mainly to understand Julia. Some of the problems are because of redefinitions of variables, e.g. https://github.com/JuliaLang/julia/blob/master/base/linalg/arnoldi.jl#L236 which I think can just be deleted. Some problems were fixed by typing the arguments to `aupd_wrapper` and `eupd_wrapper` more strictly. They are internal functions, so it should be fine with `Int` instead of `Integer`.

The main problem, though, seems to be the structuring of the wrappers. I guess that `Symmetric` should simply have its own method. This might cause some code dublication, but I think that is okay if we can get better typed return values.

It would be great if you could take a look.

**Comment by dgleich** [comment]
Thanks for the suggestion on the way to work. I'll try and take a deeper look starting with aupd_wrapper on a very simple test case.
~~~~
     A = spdiagm(1.:25.)

    @assert size(A,1) >= 20
    @assert eltype(A) == Float64
    T = eltype(A)
    
    matvecA(x) = A*x
    matvecB(x) = x
    solveSI(x) = x
    mode = 1
    n = size(A,1)
    sym = true
    iscmplx = false
    bmat = "I"
    nev = 6
    ncv = 20
    
    whichstr = "LM"
    tol = 1.e-8
    maxiter = 300
    v0 = ones(Float64,n)
    
    # Compute the Ritz values and Ritz vectors
    (resid, v, ldv, iparam, ipntr, workd, workl, lworkl, rwork, TOL) =
       TEST.aupd_wrapper(T, matvecA, matvecB, solveSI, n, sym, iscmplx, bmat, nev, ncv, whichstr, tol, maxiter, mode, v0)

    @code_warntype TEST.aupd_wrapper(T, matvecA, matvecB, solveSI, n, sym, iscmplx, bmat, nev, ncv, whichstr, tol, maxiter, mode, v0)
~~~~

This seems to mostly get things correctly typed

~~~~
... very long output ...  end::TUPLE{ANY,ARRAY{FLOAT64,2},INT64,ARRAY{INT64,1},ARRAY{INT64,1},ARRAY{FLOAT64,1},ARRAY{FLOAT64,1},INT64,ANY,ANY}
~~~~

I'll see if there's an easy way to fix up the others.

**Comment by dgleich** [comment]
A few notes... this construct appears to break the type inference method.
~~~~
    if isempty(v0)
        resid  = Array(T, n)
        info   = zeros(BlasInt, 1)
    else
        resid  = deepcopy(v0)
        info   = ones(BlasInt, 1)
    end
~~~~
It gets the correct output type if I use:
~~~~
    resid = Array(T,n)
    info = zeros(BlasInt, 1)

    # overwrite with v0 if provided
    if !isempty(v0)
        resid[:] = v0[:]
        info[1] = 1
    end
~~~~
Is there any method to handle this more elegantly? With that, there's a simple change to fix the aupd_wrapper call. I'm working on the eupd_wrapper, but that one is a little trickier. 

In chatting with a few students about this issue, we came up with an interesting solution. The raw eigenvalue/vector routines could return both real and complex solutions, where only one set is specified depending on the type of problem. For instance:
~~~~
flag,dr,Vr,dc,Vc =  eigs_arpack(A) where flag indicates whether real or complex is set.
~~~~
Of course, you'd probably want to wrap that in something more julia like using a specific eigenspace type. (Which would correspond to a subset of an Eigen type).

**Comment by dgleich** [comment]
I've got a type-stable version of eupd now. My working branch and edits are here:

https://github.com/JuliaLang/julia/compare/master...dgleich:eigs-type-stable

There is no way to make the output from eigs itself type stable because it returns one of two different tuples unfortunately (depending on the value of ritzvec in the optional parameters). This could be fixed with a more radical edit. 

As a short term fix, I'd propose adding an eigs_arpack function that is lower-level and returns both sets of eigenvalues and vectors (real and complex). This is the new output from eupd_wrapper directly. (Which is what eigs used to return.)

This would make it feasible to write type-stable code that calls this more advanced function. You'd at least be able to deterministically get real-valued output from symmetric inputs.

Right now, I'm working on making eigs_arpack type-stable. There still seems to be some issue in the code there.

**Comment by KrisDM** [comment]
I don't know if this is related, but calling eigs on a sparse Float32 matrix generates an error:

```
julia> eigs(speye(Float32,2), which=:SM, nev=1, ritzvec=false)
ERROR: StackOverflowError:
 in call at essentials.jl:56 (repeats 2 times)

julia> eigs(speye(Float64,2), which=:SM, nev=1, ritzvec=false)
([1.0],1,1,2,[0.0,0.0])
```

**Comment by ViralBShah** [comment]
@KrisDM Could you open a new issue? I don't see it on master, but it exists on 0.4.

**Comment by ViralBShah** [comment]
@dgleich Could you submit a PR with your suggestion - so that we can get something in? Should hopefully not be too much effort.

**Comment by andreasnoack** [comment]
@ViralBShah @14812 fixes the issue mentioned by @KrisDM and should probably be backported.

**Comment by KrisDM** [comment]
Does that mean I don't need to open a new issue?

**Comment by andreasnoack** [comment]
@KrisDM It shouldn't be necessary. You can try it out on 0.4 if you
```
git fetch origin
git cherry-pick 2b6c143db4bbbfd6c5213639ae86ac0c16772e99
make
```

**Comment by KrisDM** [comment]
@andreasnoack  Thanks, will give it a try.

**Comment by ViralBShah** [comment]
Will be fixed in 0.4.4 in a few weeks.

**Comment by andreasnoack** [comment]
Tracked in https://github.com/JuliaLinearAlgebra/Arpack.jl/issues/17
